SimPPS/RPDigiProducer/src/RPDisplacementGenerator.cc

```diff
@@ -0,0 +1,94 @@
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "SimDataFormats/TrackingHit/interface/PSimHit.h"
+#include "DataFormats/CTPPSDetId/interface/TotemRPDetId.h"
+#include "FWCore/Framework/interface/EventSetup.h"
+#include "FWCore/Framework/interface/ESHandle.h"
+#include "Geometry/Records/interface/VeryForwardMisalignedGeometryRecord.h"
+#include "Geometry/Records/interface/VeryForwardRealGeometryRecord.h"
+#include "Geometry/VeryForwardGeometryBuilder/interface/CTPPSGeometry.h"
+#include "CondFormats/CTPPSReadoutObjects/interface/CTPPSRPAlignmentCorrectionsData.h"
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
+#include <Math/RotationZYX.h>
+#include <Math/Rotation3D.h>
+
+#include "SimPPS/RPDigiProducer/interface/RPDisplacementGenerator.h"
+
+using namespace std;
+using namespace edm;
+
+RPDisplacementGenerator::RPDisplacementGenerator(const edm::ParameterSet &ps, RPDetId _detId, const edm::EventSetup &iSetup) : detId(_detId)
+{
+  isOn = ps.getParameter<bool>("RPDisplacementOn");
+
+  // read the alignment correction
+  ESHandle<CTPPSRPAlignmentCorrectionsData> alignments;
+  if( auto rec = iSetup.tryToGet<VeryForwardMisalignedGeometryRecord>()) {
+      iSetup.get<VeryForwardMisalignedGeometryRecord>().get(alignments);
+  } 
+
+  unsigned int decId = rawToDecId(detId);
+
+  math::XYZVectorD S_m;
+  RotationMatrix R_m;
+
+  if (alignments.isValid()) {
+    const CTPPSRPAlignmentCorrectionData& ac = alignments->getFullSensorCorrection(decId);
+    S_m = ac.getTranslation();
+    R_m = ac.getRotationMatrix();
+  } else
+    isOn = false;
+
+  // transform shift and rotation to the local coordinate frame
+  ESHandle<CTPPSGeometry> geom;
+  iSetup.get<VeryForwardRealGeometryRecord>().get(geom);
+  const DetGeomDesc *g = geom->getSensor(detId);
+  const DDRotationMatrix& R_l = g->rotation();
+  rotation = R_l.Inverse() * R_m.Inverse() * R_l;
+  shift = R_l.Inverse() * R_m.Inverse() * S_m;
+
+#ifdef DEBUG
+  edm::LogDebug("RPDisplacementGenerator")<<" det id = " << decId << ", isOn = " << isOn << "\n";
+  if (isOn) {
+    edm::LogDebug("RPDisplacementGenerator") << " shift = " << shift << "\n";
+    edm::LogDebug("RPDisplacementGenerator") << " rotation = " << rotation << "\n";
+  }
+#endif
```

Comment:
Could the checked macro be replaced with `EDM_ML_DEBUG`? Also, `LogDebug` is not in a namespace (i.e. remove `edm::`) as it is a macro. Actually how about
```cpp
LogDebug("RPDisplacementGenerator").log([&](auto& l){
  l << <" det id = " << decId << ", isOn = " << isOn << "\n";
  if (isOn) {
    l << " shift = " << shift << "\n";
    l << " rotation = " << rotation << "\n";
  }
});
```
?